// @aztec/aztec.js/dest/account/contract/ecdsa_account_contract.js
import { Ecdsa } from '@aztec/circuits.js/barretenberg';
import { AuthWitness } from '@aztec/types';
import EcdsaAccountContractArtifact from '../../artifacts/ecdsa_account_contract.json' assert { type: 'json' };
import { BaseAccountContract } from './base_account_contract.js';
/**
 * Account contract that authenticates transactions using ECDSA signatures
 * verified against a secp256k1 public key stored in an immutable encrypted note.
 */
export class EcdsaAccountContract extends BaseAccountContract {
    constructor(signingPrivateKey) {
        super(EcdsaAccountContractArtifact);
        this.signingPrivateKey = signingPrivateKey;
    }
    getDeploymentArgs() {
        const signingPublicKey = new Ecdsa().computePublicKey(this.signingPrivateKey);
        return [signingPublicKey.subarray(0, 32), signingPublicKey.subarray(32, 64)];
    }
    getAuthWitnessProvider(_address) {
        return new EcdsaAuthWitnessProvider(this.signingPrivateKey);
    }
}
/** Creates auth witnesses using ECDSA signatures. */
class EcdsaAuthWitnessProvider {
    constructor(signingPrivateKey) {
        this.signingPrivateKey = signingPrivateKey;
    }
    createAuthWitness(message) {
        const ecdsa = new Ecdsa();
        const signature = ecdsa.constructSignature(message.toBuffer(), this.signingPrivateKey);
        return Promise.resolve(new AuthWitness(message, [...signature.r, ...signature.s]));
    }
}
