// @aztec/types/dest/logs/unencrypted_l2_log.js
import { AztecAddress, FunctionSelector } from '@aztec/circuits.js';
import { BufferReader, prefixBufferWithLength } from '@aztec/foundation/serialize';
import { randomBytes } from 'crypto';
/**
 * Represents an individual unencrypted log entry.
 */
export class UnencryptedL2Log {
    constructor(
    /**
     * Address of the contract that emitted the event
     * NOTE: It would make sense to have the address only in `FunctionL2Logs` because contract address is shared for all
     * function logs. I didn't do this because it would require us to have 2 FunctionL2Logs classes (one with contract
     * address and one without) for unencrypted and encrypted because encrypted logs can't expose the address in an
     * unencrypted form. For this reason separating the classes seems like a premature optimization.
     * TODO: Optimize this once it makes sense.
     */
    contractAddress, 
    /**
     * Selector of the event/log topic.
     * TODO: https://github.com/AztecProtocol/aztec-packages/issues/2632
     */
    selector, 
    /** The data contents of the log. */
    data) {
        this.contractAddress = contractAddress;
        this.selector = selector;
        this.data = data;
    }
    get length() {
        return FunctionSelector.SIZE + this.data.length;
    }
    /**
     * Serializes log to a buffer.
     * @returns A buffer containing the serialized log.
     */
    toBuffer() {
        return Buffer.concat([
            this.contractAddress.toBuffer(),
            this.selector.toBuffer(),
            prefixBufferWithLength(this.data),
        ]);
    }
    /**
     * Serializes log to a human readable string.
     * @returns A human readable representation of the log.
     */
    toHumanReadable() {
        return `UnencryptedL2Log(contractAddress: ${this.contractAddress.toString()}, selector: ${this.selector.toString()}, data: ${this.data.toString('ascii')})`;
    }
    /**
     * Deserializes log from a buffer.
     * @param buffer - The buffer or buffer reader containing the log.
     * @returns Deserialized instance of `Log`.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const contractAddress = AztecAddress.fromBuffer(reader);
        const selector = FunctionSelector.fromBuffer(reader);
        const data = reader.readBuffer();
        return new UnencryptedL2Log(contractAddress, selector, data);
    }
    /**
     * Crates a random log.
     * @returns A random log.
     */
    static random() {
        const contractAddress = AztecAddress.random();
        const selector = new FunctionSelector(Math.floor(Math.random() * (2 ** (FunctionSelector.SIZE * 8) - 1)));
        const dataLength = FunctionSelector.SIZE + Math.floor(Math.random() * 200);
        const data = randomBytes(dataLength);
        return new UnencryptedL2Log(contractAddress, selector, data);
    }
}
