// @aztec/noir-contracts/dest/types/docs_example.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EthAddress, Point, } from '@aztec/aztec.js';
import DocsExampleContractArtifactJson from '../artifacts/docs_example_contract.json' assert { type: 'json' };
export const DocsExampleContractArtifact = DocsExampleContractArtifactJson;
/**
 * Type-safe interface for contract DocsExample;
 */
export class DocsExampleContract extends ContractBase {
    constructor(completeAddress, wallet, portalContract = EthAddress.ZERO) {
        super(completeAddress, DocsExampleContractArtifact, wallet, portalContract);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, DocsExampleContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, min_points, max_points, legendary_card_secret) {
        return new DeployMethod(Point.ZERO, wallet, DocsExampleContractArtifact, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, min_points, max_points, legendary_card_secret) {
        return new DeployMethod(publicKey, wallet, DocsExampleContractArtifact, Array.from(arguments).slice(2));
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return DocsExampleContractArtifact;
    }
}
