// @aztec/noir-contracts/dest/types/inclusion_proofs.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EthAddress, Point, } from '@aztec/aztec.js';
import InclusionProofsContractArtifactJson from '../artifacts/inclusion_proofs_contract.json' assert { type: 'json' };
export const InclusionProofsContractArtifact = InclusionProofsContractArtifactJson;
/**
 * Type-safe interface for contract InclusionProofs;
 */
export class InclusionProofsContract extends ContractBase {
    constructor(completeAddress, wallet, portalContract = EthAddress.ZERO) {
        super(completeAddress, InclusionProofsContractArtifact, wallet, portalContract);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, InclusionProofsContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, public_value) {
        return new DeployMethod(Point.ZERO, wallet, InclusionProofsContractArtifact, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, public_value) {
        return new DeployMethod(publicKey, wallet, InclusionProofsContractArtifact, Array.from(arguments).slice(2));
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return InclusionProofsContractArtifact;
    }
}
