// @aztec/noir-contracts/dest/types/schnorr_single_key_account.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EthAddress, Point, } from '@aztec/aztec.js';
import SchnorrSingleKeyAccountContractArtifactJson from '../artifacts/schnorr_single_key_account_contract.json' assert { type: 'json' };
export const SchnorrSingleKeyAccountContractArtifact = SchnorrSingleKeyAccountContractArtifactJson;
/**
 * Type-safe interface for contract SchnorrSingleKeyAccount;
 */
export class SchnorrSingleKeyAccountContract extends ContractBase {
    constructor(completeAddress, wallet, portalContract = EthAddress.ZERO) {
        super(completeAddress, SchnorrSingleKeyAccountContractArtifact, wallet, portalContract);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, SchnorrSingleKeyAccountContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(Point.ZERO, wallet, SchnorrSingleKeyAccountContractArtifact, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet) {
        return new DeployMethod(publicKey, wallet, SchnorrSingleKeyAccountContractArtifact, Array.from(arguments).slice(2));
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return SchnorrSingleKeyAccountContractArtifact;
    }
}
