// @aztec/noir-contracts/dest/types/token_blacklist.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EthAddress, Point, } from '@aztec/aztec.js';
import TokenBlacklistContractArtifactJson from '../artifacts/token_blacklist_contract.json' assert { type: 'json' };
export const TokenBlacklistContractArtifact = TokenBlacklistContractArtifactJson;
/**
 * Type-safe interface for contract TokenBlacklist;
 */
export class TokenBlacklistContract extends ContractBase {
    constructor(completeAddress, wallet, portalContract = EthAddress.ZERO) {
        super(completeAddress, TokenBlacklistContractArtifact, wallet, portalContract);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, TokenBlacklistContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, admin, slow_updates_contract) {
        return new DeployMethod(Point.ZERO, wallet, TokenBlacklistContractArtifact, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, admin, slow_updates_contract) {
        return new DeployMethod(publicKey, wallet, TokenBlacklistContractArtifact, Array.from(arguments).slice(2));
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return TokenBlacklistContractArtifact;
    }
}
