// @aztec/foundation/dest/log/logger.js
import debug from 'debug';
import isNode from 'detect-node';
import { isatty } from 'tty';
// Matches a subset of Winston log levels
const LogLevels = ['silent', 'error', 'warn', 'info', 'verbose', 'debug'];
const DefaultLogLevel = process.env.NODE_ENV === 'test' ? 'silent' : 'info';
const envLogLevel = process.env.LOG_LEVEL?.toLowerCase();
const currentLevel = LogLevels.includes(envLogLevel) ? envLogLevel : DefaultLogLevel;
/**
 * Creates a new DebugLogger for the current module, defaulting to the LOG_LEVEL env var.
 * If DEBUG="[module]" env is set, will enable debug logging if the module matches.
 * Uses npm debug for debug level and console.error for other levels.
 * @param name - Name of the module.
 * @returns A debug logger.
 */
export function createDebugLogger(name) {
    const debugLogger = debug(name);
    if (currentLevel === 'debug') {
        debugLogger.enabled = true;
    }
    const logger = {
        silent: () => { },
        error: (msg, err, data) => logWithDebug(debugLogger, 'error', fmtErr(msg, err), data),
        warn: (msg, data) => logWithDebug(debugLogger, 'warn', msg, data),
        info: (msg, data) => logWithDebug(debugLogger, 'info', msg, data),
        verbose: (msg, data) => logWithDebug(debugLogger, 'verbose', msg, data),
        debug: (msg, data) => logWithDebug(debugLogger, 'debug', msg, data),
    };
    return Object.assign((msg, data) => logWithDebug(debugLogger, 'debug', msg, data), logger);
}
const logHandlers = [];
/**
 * Registers a callback for all logs, whether they are emitted in the current log level or not.
 * @param handler - Callback to be called on every log.
 */
export function onLog(handler) {
    logHandlers.push(handler);
}
/**
 * Logs args to npm debug if enabled or log level is debug, console.error otherwise.
 * @param debug - Instance of npm debug.
 * @param level - Intended log level.
 * @param args - Args to log.
 */
function logWithDebug(debug, level, msg, data) {
    for (const handler of logHandlers) {
        handler(level, debug.namespace, msg, data);
    }
    msg = data ? `${msg} ${fmtLogData(data)}` : msg;
    if (debug.enabled) {
        if (level !== 'debug') {
            msg = `${level.toUpperCase()} ${msg}`;
        }
        debug(msg);
    }
    else if (LogLevels.indexOf(level) <= LogLevels.indexOf(currentLevel)) {
        printLog(`${getPrefix(debug, level)} ${msg}`);
    }
}
/**
 * Returns a log prefix that emulates that of npm debug. Uses colors if in node and in a tty.
 * @param debugLogger - Instance of npm debug logger.
 * @param level - Intended log level (printed out if strictly above current log level).
 * @returns Log prefix.
 */
function getPrefix(debugLogger, level) {
    const levelLabel = currentLevel !== level ? ` ${level.toUpperCase()}` : '';
    const prefix = `${debugLogger.namespace.replace(/^aztec:/, '')}${levelLabel}`;
    if (!isNode || !isatty(process.stderr.fd)) {
        return prefix;
    }
    const colorIndex = debug.selectColor(debugLogger.namespace);
    const colorCode = '\u001B[3' + (colorIndex < 8 ? colorIndex : '8;5;' + colorIndex);
    return `  ${colorCode};1m${prefix}\u001B[0m`;
}
/**
 * Outputs to console error.
 * @param msg - What to log.
 */
function printLog(msg) {
    // eslint-disable-next-line no-console
    console.error(msg);
}
/**
 * Concatenates a log message and an exception.
 * @param msg - Log message
 * @param err - Error to log
 * @returns A string with both the log message and the error message.
 */
function fmtErr(msg, err) {
    const errStr = err && [err.name, err.message].filter(x => !!x).join(' ');
    return err ? `${msg}: ${errStr || err}` : msg;
}
/**
 * Formats structured log data as a string for console output.
 * @param data - Optional log data.
 */
function fmtLogData(data) {
    return Object.entries(data ?? {})
        .map(([key, value]) => `${key}=${value}`)
        .join(' ');
}
