// @aztec/noir-contracts/dest/artifacts/index.js
import BenchmarkingJson from './benchmarking_contract.json' assert { type: 'json' };
export const BenchmarkingContractArtifact = BenchmarkingJson;
import CardGameJson from './card_game_contract.json' assert { type: 'json' };
export const CardGameContractArtifact = CardGameJson;
import ChildJson from './child_contract.json' assert { type: 'json' };
export const ChildContractArtifact = ChildJson;
import CounterJson from './counter_contract.json' assert { type: 'json' };
export const CounterContractArtifact = CounterJson;
import DocsExampleJson from './docs_example_contract.json' assert { type: 'json' };
export const DocsExampleContractArtifact = DocsExampleJson;
import EasyPrivateTokenJson from './easy_private_token_contract.json' assert { type: 'json' };
export const EasyPrivateTokenContractArtifact = EasyPrivateTokenJson;
import EasyPrivateVotingJson from './easy_private_voting_contract.json' assert { type: 'json' };
export const EasyPrivateVotingContractArtifact = EasyPrivateVotingJson;
import EcdsaAccountJson from './ecdsa_account_contract.json' assert { type: 'json' };
export const EcdsaAccountContractArtifact = EcdsaAccountJson;
import EscrowJson from './escrow_contract.json' assert { type: 'json' };
export const EscrowContractArtifact = EscrowJson;
import ImportTestJson from './import_test_contract.json' assert { type: 'json' };
export const ImportTestContractArtifact = ImportTestJson;
import InclusionProofsJson from './inclusion_proofs_contract.json' assert { type: 'json' };
export const InclusionProofsContractArtifact = InclusionProofsJson;
import LendingJson from './lending_contract.json' assert { type: 'json' };
export const LendingContractArtifact = LendingJson;
import ParentJson from './parent_contract.json' assert { type: 'json' };
export const ParentContractArtifact = ParentJson;
import PendingCommitmentsJson from './pending_commitments_contract.json' assert { type: 'json' };
export const PendingCommitmentsContractArtifact = PendingCommitmentsJson;
import PriceFeedJson from './price_feed_contract.json' assert { type: 'json' };
export const PriceFeedContractArtifact = PriceFeedJson;
import SchnorrAccountJson from './schnorr_account_contract.json' assert { type: 'json' };
export const SchnorrAccountContractArtifact = SchnorrAccountJson;
import SchnorrHardcodedAccountJson from './schnorr_hardcoded_account_contract.json' assert { type: 'json' };
export const SchnorrHardcodedAccountContractArtifact = SchnorrHardcodedAccountJson;
import SchnorrSingleKeyAccountJson from './schnorr_single_key_account_contract.json' assert { type: 'json' };
export const SchnorrSingleKeyAccountContractArtifact = SchnorrSingleKeyAccountJson;
import SlowTreeJson from './slow_tree_contract.json' assert { type: 'json' };
export const SlowTreeContractArtifact = SlowTreeJson;
import StatefulTestJson from './stateful_test_contract.json' assert { type: 'json' };
export const StatefulTestContractArtifact = StatefulTestJson;
import TestJson from './test_contract.json' assert { type: 'json' };
export const TestContractArtifact = TestJson;
import TokenBlacklistJson from './token_blacklist_contract.json' assert { type: 'json' };
export const TokenBlacklistContractArtifact = TokenBlacklistJson;
import TokenBridgeJson from './token_bridge_contract.json' assert { type: 'json' };
export const TokenBridgeContractArtifact = TokenBridgeJson;
import TokenJson from './token_contract.json' assert { type: 'json' };
export const TokenContractArtifact = TokenJson;
import UniswapJson from './uniswap_contract.json' assert { type: 'json' };
export const UniswapContractArtifact = UniswapJson;
