// @aztec/foundation/dest/transport/browser/shared_worker_listener.js
import EventEmitter from 'events';
import { MessagePortSocket } from './message_port_socket.js';
/**
 * SharedWorkerListener is an extension of the EventEmitter class that implements the Listener interface.
 * It provides functionality to handle incoming messages from a shared worker and emit events for new sockets
 * created in response to these incoming connections. This class is meant to be used in the context of managing
 * MessagePort connections within the SharedWorkerGlobalScope.
 */
export class SharedWorkerListener extends EventEmitter {
    constructor(worker) {
        super();
        this.worker = worker;
        this.handleMessageEvent = (event) => {
            const [port] = event.ports;
            if (!port) {
                return;
            }
            this.emit('new_socket', new MessagePortSocket(port));
        };
    }
    /**
     * Initializes the shared worker listener by assigning the 'handleMessageEvent' method as the event handler
     * for the 'onconnect' event of the SharedWorkerGlobalScope. The 'handleMessageEvent' function will be called
     * whenever a new connection is established with the shared worker.
     */
    open() {
        this.worker.onconnect = this.handleMessageEvent;
    }
    /**
     * Closes the SharedWorkerListener by detaching the 'onconnect' event handler.
     * This stops the listener from emitting new sockets on incoming connections.
     */
    close() {
        this.worker.onconnect = () => { };
    }
}
