// @aztec/foundation/dest/transport/browser/worker_listener.js
import EventEmitter from 'events';
import { MessagePortSocket } from './message_port_socket.js';
/**
 * WorkerListener is a class that extends EventEmitter and implements the Listener interface.
 * It listens for incoming connections on a dedicated worker global scope, and emits a 'new_socket' event
 * with a MessagePortSocket instance for each new connection. This allows applications to communicate
 * with other workers or main thread through the MessagePortSocket abstraction.
 *
 * The open() method starts listening for incoming connections, while the close() method stops it.
 */
export class WorkerListener extends EventEmitter {
    constructor(worker) {
        super();
        this.worker = worker;
        this.handleMessageEvent = (event) => {
            const [port] = event.ports;
            if (!port) {
                return;
            }
            this.emit('new_socket', new MessagePortSocket(port));
        };
    }
    /**
     * Initializes the WorkerListener by setting the 'onmessage' event handler of the worker.
     * The 'onmessage' event will be triggered when the worker receives a message, and it will then
     * call the handleMessageEvent method to handle incoming connections.
     */
    open() {
        this.worker.onmessage = this.handleMessageEvent;
    }
    /**
     * Close the worker listener by removing the 'onmessage' event handler.
     * This method effectively stops the WorkerListener from reacting to new incoming messages.
     */
    close() {
        this.worker.onmessage = () => { };
    }
}
