// @aztec/circuits.js/dest/structs/aggregation_object.js
import { Fq, Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import times from 'lodash.times';
import { serializeToBuffer } from '../utils/serialize.js';
import { Vector } from './shared.js';
import { G1AffineElement } from './verification_key.js';
/**
 * Contains the aggregated proof of all the previous kernel iterations.
 *
 * See barretenberg/cpp/src/barretenberg/stdlib/recursion/aggregation_state/native_aggregation_state.hpp
 * for more context.
 */
export class AggregationObject {
    constructor(
    /**
     * One of the 2 aggregated elements storing the verification results of proofs in the past.
     */
    p0, 
    /**
     * One of the 2 aggregated elements storing the verification results of proofs in the past.
     */
    p1, 
    /**
     * The public inputs of the inner proof (these become the private inputs to the recursive circuit).
     */
    publicInputs, 
    /**
     * Witness indices that point to (P0, P1).
     */
    proofWitnessIndices, 
    /**
     * Indicates if this aggregation state contain past (P0, P1).
     */
    hasData = false) {
        this.p0 = p0;
        this.p1 = p1;
        this.publicInputs = publicInputs;
        this.proofWitnessIndices = proofWitnessIndices;
        this.hasData = hasData;
    }
    toBuffer() {
        return serializeToBuffer(this.p0, this.p1, new Vector(this.publicInputs), new Vector(this.proofWitnessIndices), this.hasData);
    }
    /**
     * Deserializes this object from a buffer.
     * @param buffer - The buffer representation of this object.
     * @returns The deserialized object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new AggregationObject(reader.readObject(G1AffineElement), reader.readObject(G1AffineElement), reader.readVector(Fr), reader.readNumberVector(), reader.readBoolean());
    }
    /**
     * Creates a fake object for testing.
     * @returns The fake object.
     */
    static makeFake() {
        return new AggregationObject(new G1AffineElement(new Fq(1n), new Fq(2n)), new G1AffineElement(new Fq(1n), new Fq(2n)), [], times(16, i => 3027 + i), false);
    }
}
