// @aztec/circuits.js/dest/structs/call_context.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { EthAddress } from '@aztec/foundation/eth-address';
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../utils/serialize.js';
import { Fr, FunctionSelector } from './index.js';
/**
 * Call context.
 * @see abis/call_context.hpp
 */
export class CallContext {
    constructor(
    /**
     * Address of the account which represents the entity who invoked the call.
     */
    msgSender, 
    /**
     * The contract address against which all state changes will be stored. Not called `contractAddress` because during
     * delegate call the contract whose code is being executed may be different from the contract whose state is being
     * modified.
     */
    storageContractAddress, 
    /**
     * Address of the portal contract to the storage contract.
     * Union type is a kludge until C++ has an eth address type.
     */
    portalContractAddress, 
    /**
     * Function selector of the function being called.
     */
    functionSelector, 
    /**
     * Determines whether the call is a delegate call (see Ethereum's delegate call opcode for more information).
     */
    isDelegateCall, 
    /**
     * Determines whether the call is modifying state.
     */
    isStaticCall, 
    /**
     * Determines whether the call is a contract deployment.
     */
    isContractDeployment) {
        this.msgSender = msgSender;
        this.storageContractAddress = storageContractAddress;
        this.functionSelector = functionSelector;
        this.isDelegateCall = isDelegateCall;
        this.isStaticCall = isStaticCall;
        this.isContractDeployment = isContractDeployment;
        this.portalContractAddress =
            portalContractAddress instanceof EthAddress ? portalContractAddress : EthAddress.fromField(portalContractAddress);
    }
    /**
     * Returns a new instance of CallContext with zero msg sender, storage contract address and portal contract address.
     * @returns A new instance of CallContext with zero msg sender, storage contract address and portal contract address.
     */
    static empty() {
        return new CallContext(AztecAddress.ZERO, AztecAddress.ZERO, Fr.ZERO, FunctionSelector.empty(), false, false, false);
    }
    isEmpty() {
        return (this.msgSender.isZero() &&
            this.storageContractAddress.isZero() &&
            this.portalContractAddress.isZero() &&
            this.functionSelector.isEmpty());
    }
    static from(fields) {
        return new CallContext(...CallContext.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.msgSender,
            fields.storageContractAddress,
            fields.portalContractAddress,
            fields.functionSelector,
            fields.isDelegateCall,
            fields.isStaticCall,
            fields.isContractDeployment,
        ];
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...CallContext.getFields(this));
    }
    /**
     * Deserialize this from a buffer.
     * @param buffer - The bufferable type from which to deserialize.
     * @returns The deserialized instance of PublicCallRequest.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new CallContext(new AztecAddress(reader.readBytes(32)), new AztecAddress(reader.readBytes(32)), new EthAddress(reader.readBytes(32)), FunctionSelector.fromBuffer(reader.readBytes(4)), reader.readBoolean(), reader.readBoolean(), reader.readBoolean());
    }
    equals(callContext) {
        return (callContext.msgSender.equals(this.msgSender) &&
            callContext.storageContractAddress.equals(this.storageContractAddress) &&
            callContext.portalContractAddress.equals(this.portalContractAddress) &&
            callContext.functionSelector.equals(this.functionSelector) &&
            callContext.isDelegateCall === this.isDelegateCall &&
            callContext.isStaticCall === this.isStaticCall &&
            callContext.isContractDeployment === this.isContractDeployment);
    }
}
