// @aztec/circuits.js/dest/structs/call_request.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../utils/serialize.js';
/**
 * Caller context.
 */
export class CallerContext {
    constructor(
    /**
     * Address of the caller contract.
     */
    msgSender, 
    /**
     * Storage contract address of the caller contract.
     */
    storageContractAddress) {
        this.msgSender = msgSender;
        this.storageContractAddress = storageContractAddress;
    }
    /**
     * Returns a new instance of CallerContext with zero values.
     * @returns A new instance of CallerContext with zero values.
     */
    static empty() {
        return new CallerContext(AztecAddress.ZERO, AztecAddress.ZERO);
    }
    isEmpty() {
        return this.msgSender.isZero() && this.storageContractAddress.isZero();
    }
    static from(fields) {
        return new CallerContext(...CallerContext.getFields(fields));
    }
    static getFields(fields) {
        return [fields.msgSender, fields.storageContractAddress];
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...CallerContext.getFields(this));
    }
    /**
     * Deserialize this from a buffer.
     * @param buffer - The bufferable type from which to deserialize.
     * @returns The deserialized instance of PublicCallRequest.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new CallerContext(new AztecAddress(reader.readBytes(32)), new AztecAddress(reader.readBytes(32)));
    }
    equals(callerContext) {
        return (callerContext.msgSender.equals(this.msgSender) &&
            callerContext.storageContractAddress.equals(this.storageContractAddress));
    }
}
/**
 * Call request.
 */
export class CallRequest {
    constructor(
    /**
     * The hash of the call stack item.
     */
    hash, 
    /**
     * The address of the contract calling the function.
     */
    callerContractAddress, 
    /**
     * The call context of the contract calling the function.
     */
    callerContext) {
        this.hash = hash;
        this.callerContractAddress = callerContractAddress;
        this.callerContext = callerContext;
    }
    toBuffer() {
        return serializeToBuffer(this.hash, this.callerContractAddress, this.callerContext);
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance of CallRequest.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new CallRequest(Fr.fromBuffer(reader), reader.readObject(AztecAddress), reader.readObject(CallerContext));
    }
    isEmpty() {
        return this.hash.isZero() && this.callerContractAddress.isZero() && this.callerContext.isEmpty();
    }
    /**
     * Returns a new instance of CallRequest with zero hash, caller contract address and caller context.
     * @returns A new instance of CallRequest with zero hash, caller contract address and caller context.
     */
    static empty() {
        return new CallRequest(Fr.ZERO, AztecAddress.ZERO, CallerContext.empty());
    }
    equals(callRequest) {
        return (callRequest.hash.equals(this.hash) &&
            callRequest.callerContractAddress.equals(this.callerContractAddress) &&
            callRequest.callerContext.equals(this.callerContext));
    }
}
