// @aztec/circuits.js/dest/structs/membership_witness.js
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { assertMemberLength, range } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
/**
 * Contains information which can be used to prove that a leaf is a member of a Merkle tree.
 */
export class MembershipWitness {
    constructor(
    /**
     * Size of the sibling path (number of fields it contains).
     */
    pathSize, 
    /**
     * Index of a leaf in the Merkle tree.
     */
    leafIndex, 
    /**
     * Sibling path of the leaf in the Merkle tree.
     */
    siblingPath) {
        this.leafIndex = leafIndex;
        this.siblingPath = siblingPath;
        assertMemberLength(this, 'siblingPath', pathSize);
    }
    toBuffer() {
        return serializeToBuffer(toBufferBE(this.leafIndex, 32), ...this.siblingPath);
    }
    static mock(size, start) {
        return new MembershipWitness(size, BigInt(start), range(size, start).map(x => new Fr(BigInt(x))));
    }
    /**
     * Creates a random membership witness. Used for testing purposes.
     * @param pathSize - Number of fields in the sibling path.
     * @returns Random membership witness.
     */
    static random(pathSize) {
        return new MembershipWitness(pathSize, 0n, Array(pathSize)
            .fill(0)
            .map(() => Fr.random()));
    }
    /**
     * Creates a membership witness whose sibling path is full of zero fields.
     * @param pathSize - Number of fields in the sibling path.
     * @param leafIndex - Index of the leaf in the Merkle tree.
     * @returns Membership witness with zero sibling path.
     */
    static empty(pathSize, leafIndex) {
        const arr = Array(pathSize)
            .fill(0)
            .map(() => Fr.ZERO);
        return new MembershipWitness(pathSize, leafIndex, arr);
    }
    static fromBufferArray(leafIndex, siblingPath) {
        return new MembershipWitness(siblingPath.length, leafIndex, siblingPath.map(x => Fr.fromBuffer(x)));
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized `MembershipWitness`.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const leafIndex = toBigIntBE(reader.readBytes(32));
        const siblingPath = reader.readBufferArray();
        return this.fromBufferArray(leafIndex, siblingPath);
    }
    /**
     * Creates a deserializer object for a MembershipWitness with a given size.
     * @param size - Expected size of the witness.
     * @returns A deserializer object.
     */
    static deserializer(size) {
        return {
            fromBuffer: (buffer) => {
                const reader = BufferReader.asReader(buffer);
                const leafIndex = toBigIntBE(reader.readBytes(32));
                const siblingPath = reader.readArray(size, Fr);
                return new MembershipWitness(size, leafIndex, siblingPath);
            },
        };
    }
}
