// @aztec/circuits.js/dest/structs/read_request_membership_witness.js
import { toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { MAX_NEW_COMMITMENTS_PER_CALL, NOTE_HASH_TREE_HEIGHT } from '../constants.gen.js';
import { makeTuple, range } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
/**
 * A ReadRequestMembershipWitness is similar to a MembershipWitness but includes
 * some additional fields used to direct the kernel regarding whether a read is transient
 * and if so which commitment it corresponds to.
 */
export class ReadRequestMembershipWitness {
    constructor(
    /**
     * Index of a leaf in the Merkle tree.
     */
    leafIndex, 
    /**
     * Sibling path of the leaf in the Merkle tree.
     */
    siblingPath, 
    /**
     * Whether or not the read request corresponds to a pending commitment.
     */
    isTransient = false, 
    /**
     * When transient, the commitment being read was created by some app circuit in the current TX.
     * The kernel will need some hint to efficiently find that commitment for a given read request.
     * When not transient, this can be 0.
     */
    hintToCommitment) {
        this.leafIndex = leafIndex;
        this.siblingPath = siblingPath;
        this.isTransient = isTransient;
        this.hintToCommitment = hintToCommitment;
        if (hintToCommitment.toBigInt() > MAX_NEW_COMMITMENTS_PER_CALL) {
            throw new Error(`Expected ReadRequestMembershipWitness' hintToCommitment(${hintToCommitment}) to be <= NEW_COMMITMENTS_LENGTH(${MAX_NEW_COMMITMENTS_PER_CALL})`);
        }
    }
    toBuffer() {
        return serializeToBuffer(toBufferBE(this.leafIndex.toBigInt(), 32), ...this.siblingPath, this.isTransient, this.hintToCommitment);
    }
    static mock(size, start) {
        return new ReadRequestMembershipWitness(new Fr(start), range(size, start).map(x => new Fr(BigInt(x))), false, new Fr(0));
    }
    /**
     * Creates a random membership witness. Used for testing purposes.
     * @returns Random membership witness.
     */
    static random() {
        return new ReadRequestMembershipWitness(new Fr(0n), makeTuple(NOTE_HASH_TREE_HEIGHT, () => Fr.random()), false, new Fr(0));
    }
    /**
     * Creates a read request membership witness whose sibling path is full of zero fields.
     * @param leafIndex - Index of the leaf in the Merkle tree.
     * @returns Membership witness with zero sibling path.
     */
    static empty(leafIndex) {
        const arr = makeTuple(NOTE_HASH_TREE_HEIGHT, () => Fr.ZERO);
        return new ReadRequestMembershipWitness(new Fr(leafIndex), arr, false, new Fr(0));
    }
    /**
     * Creates a transient read request membership witness.
     * @returns an empty transient read request membership witness.
     */
    static emptyTransient() {
        const arr = makeTuple(NOTE_HASH_TREE_HEIGHT, () => Fr.ZERO);
        return new ReadRequestMembershipWitness(new Fr(0), arr, true, new Fr(0));
    }
    static fromBufferArray(leafIndex, siblingPath, isTransient, hintToCommitment) {
        return new ReadRequestMembershipWitness(leafIndex, siblingPath.map(x => Fr.fromBuffer(x)), isTransient, hintToCommitment);
    }
    static fromMembershipWitness(membershipWitness, isTransient, hintToCommitment) {
        return new ReadRequestMembershipWitness(new Fr(membershipWitness.leafIndex), membershipWitness.siblingPath, isTransient, hintToCommitment);
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized `ReadRequestMembershipWitness`.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const leafIndex = Fr.fromBuffer(reader);
        const siblingPath = reader.readArray(NOTE_HASH_TREE_HEIGHT, Fr);
        const isTransient = reader.readBoolean();
        const hintToCommitment = Fr.fromBuffer(reader);
        return new ReadRequestMembershipWitness(leafIndex, siblingPath, isTransient, hintToCommitment);
    }
}
