// @aztec/circuits.js/dest/structs/tx_context.js
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../utils/serialize.js';
import { EthAddress, Fr, Point } from './index.js';
/**
 * Contract deployment data in a TxContext
 * cpp/src/aztec3/circuits/abis/contract_deployment_data.hpp.
 *
 * Not to be confused with NewContractData.
 */
export class ContractDeploymentData {
    constructor(
    /** Public key of the contract deployer (used when deploying account contracts). */
    deployerPublicKey, 
    /** Hash of the constructor verification key. */
    constructorVkHash, 
    /** Function tree root. */
    functionTreeRoot, 
    /** Contract address salt (used when deriving a contract address). */
    contractAddressSalt, 
    /**
     * Ethereum address of the portal contract on L1.
     * TODO(AD): union type kludge due to cbind compiler having special needs
     */
    portalContractAddress) {
        this.deployerPublicKey = deployerPublicKey;
        this.constructorVkHash = constructorVkHash;
        this.functionTreeRoot = functionTreeRoot;
        this.contractAddressSalt = contractAddressSalt;
        this.portalContractAddress = EthAddress.fromField(portalContractAddress.toField());
    }
    toBuffer() {
        return serializeToBuffer(this.deployerPublicKey, this.constructorVkHash, this.functionTreeRoot, this.contractAddressSalt, this.portalContractAddress);
    }
    /**
     * Returns an empty ContractDeploymentData.
     * @returns The empty ContractDeploymentData.
     */
    static empty() {
        return new ContractDeploymentData(Point.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, EthAddress.ZERO);
    }
    isEmpty() {
        return (this.deployerPublicKey.isZero() &&
            this.constructorVkHash.isZero() &&
            this.functionTreeRoot.isZero() &&
            this.contractAddressSalt.isZero() &&
            this.portalContractAddress.isZero());
    }
    /**
     * Deserializes contract deployment data rom a buffer or reader.
     * @param buffer - Buffer to read from.
     * @returns The deserialized ContractDeploymentData.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ContractDeploymentData(reader.readObject(Point), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), new EthAddress(reader.readBytes(32)));
    }
}
/**
 * Transaction context.
 * @see cpp/src/aztec3/circuits/abis/tx_context.hpp.
 */
export class TxContext {
    constructor(
    /**
     * Whether this is a fee paying tx. If not other tx in a bundle will pay the fee.
     * TODO(#3417): Remove fee and rebate payment fields.
     */
    isFeePaymentTx, 
    /**
     * Indicates whether this a gas rebate payment tx.
     *
     * NOTE: The following is a WIP and it is likely to change in the future.
     * Explanation: Each tx is actually 3 txs in one: a fee-paying tx, the actual tx you want to execute, and a rebate
     * tx. The fee-paying tx pays some `max_fee = gas_price * gas_limit`. Then the actual tx will cost an amount of gas
     * to execute (actual_fee = gas_price * gas_used). Then the rebate tx returns `max_fee - actual_fee` back to
     * the user.
     */
    isRebatePaymentTx, 
    /**
     * Whether this is a contract deployment tx.
     */
    isContractDeploymentTx, 
    /**
     * Contract deployment data.
     */
    contractDeploymentData, 
    /**
     * Chain ID of the transaction. Here for replay protection.
     */
    chainId, 
    /**
     * Version of the transaction. Here for replay protection.
     */
    version) {
        this.isFeePaymentTx = isFeePaymentTx;
        this.isRebatePaymentTx = isRebatePaymentTx;
        this.isContractDeploymentTx = isContractDeploymentTx;
        this.contractDeploymentData = contractDeploymentData;
        this.chainId = chainId;
        this.version = version;
    }
    /**
     * Serialize as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.isFeePaymentTx, this.isRebatePaymentTx, this.isContractDeploymentTx, this.contractDeploymentData, this.chainId, this.version);
    }
    /**
     * Deserializes TxContext from a buffer or reader.
     * @param buffer - Buffer to read from.
     * @returns The TxContext.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new TxContext(reader.readBoolean(), reader.readBoolean(), reader.readBoolean(), reader.readObject(ContractDeploymentData), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    static empty(chainId = 0, version = 0) {
        return new TxContext(false, false, false, ContractDeploymentData.empty(), new Fr(chainId), new Fr(version));
    }
    isEmpty() {
        return (!this.isFeePaymentTx &&
            !this.isRebatePaymentTx &&
            !this.isContractDeploymentTx &&
            this.contractDeploymentData.isEmpty() &&
            this.chainId.isZero() &&
            this.version.isZero());
    }
    /**
     * Create a new instance from a fields dictionary.
     * @param fields - The dictionary.
     * @returns A new instance.
     */
    static from(fields) {
        return new TxContext(...TxContext.getFields(fields));
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            fields.isFeePaymentTx,
            fields.isRebatePaymentTx,
            fields.isContractDeploymentTx,
            fields.contractDeploymentData,
            fields.chainId,
            fields.version,
        ];
    }
}
