// @aztec/aztec.js/dest/wallet/account_wallet.js
import { Fr } from '@aztec/circuits.js';
import { ContractFunctionInteraction } from '../index.js';
import { BaseWallet } from './base_wallet.js';
/**
 * A wallet implementation that forwards authentication requests to a provided account.
 */
export class AccountWallet extends BaseWallet {
    constructor(pxe, account) {
        super(pxe);
        this.account = account;
    }
    createTxExecutionRequest(execs) {
        return this.account.createTxExecutionRequest(execs);
    }
    async createAuthWitness(message) {
        message = Buffer.isBuffer(message) ? Fr.fromBuffer(message) : message;
        const witness = await this.account.createAuthWitness(message);
        await this.pxe.addAuthWitness(witness);
        return witness;
    }
    /**
     * Returns a function interaction to set a message hash as authorized in this account.
     * Public calls can then consume this authorization.
     * @param message - Message hash to authorize.
     * @param authorized - True to authorize, false to revoke authorization.
     * @returns - A function interaction.
     */
    setPublicAuth(message, authorized) {
        const args = [message, authorized];
        return new ContractFunctionInteraction(this, this.getAddress(), this.getSetIsValidStorageAbi(), args);
    }
    /** Returns the complete address of the account that implements this wallet. */
    getCompleteAddress() {
        return this.account.getCompleteAddress();
    }
    /** Returns the address of the account that implements this wallet. */
    getAddress() {
        return this.getCompleteAddress().address;
    }
    getSetIsValidStorageAbi() {
        return {
            name: 'set_is_valid_storage',
            functionType: 'open',
            isInternal: true,
            parameters: [
                {
                    name: 'message_hash',
                    type: { kind: 'field' },
                    visibility: 'private',
                },
                {
                    name: 'value',
                    type: { kind: 'boolean' },
                    visibility: 'private',
                },
            ],
            returnTypes: [],
        };
    }
}
/**
 * Extends {@link AccountWallet} with the encryption private key. Not required for
 * implementing the wallet interface but useful for testing purposes or exporting
 * an account to another pxe.
 */
export class AccountWalletWithPrivateKey extends AccountWallet {
    constructor(pxe, account, encryptionPrivateKey) {
        super(pxe, account);
        this.encryptionPrivateKey = encryptionPrivateKey;
    }
    /** Returns the encryption private key associated with this account. */
    getEncryptionPrivateKey() {
        return this.encryptionPrivateKey;
    }
}
