// @aztec/aztec.js/dest/wallet/base_wallet.js
/**
 * A base class for Wallet implementations
 */
export class BaseWallet {
    constructor(pxe) {
        this.pxe = pxe;
    }
    addCapsule(capsule) {
        return this.pxe.addCapsule(capsule);
    }
    registerAccount(privKey, partialAddress) {
        return this.pxe.registerAccount(privKey, partialAddress);
    }
    registerRecipient(account) {
        return this.pxe.registerRecipient(account);
    }
    getRegisteredAccounts() {
        return this.pxe.getRegisteredAccounts();
    }
    getRegisteredAccount(address) {
        return this.pxe.getRegisteredAccount(address);
    }
    getRecipients() {
        return this.pxe.getRecipients();
    }
    getRecipient(address) {
        return this.pxe.getRecipient(address);
    }
    addContracts(contracts) {
        return this.pxe.addContracts(contracts);
    }
    getContracts() {
        return this.pxe.getContracts();
    }
    simulateTx(txRequest, simulatePublic) {
        return this.pxe.simulateTx(txRequest, simulatePublic);
    }
    sendTx(tx) {
        return this.pxe.sendTx(tx);
    }
    getTx(txHash) {
        return this.pxe.getTx(txHash);
    }
    getTxReceipt(txHash) {
        return this.pxe.getTxReceipt(txHash);
    }
    getNotes(filter) {
        return this.pxe.getNotes(filter);
    }
    getPublicStorageAt(contract, storageSlot) {
        return this.pxe.getPublicStorageAt(contract, storageSlot);
    }
    addNote(note) {
        return this.pxe.addNote(note);
    }
    getBlock(number) {
        return this.pxe.getBlock(number);
    }
    viewTx(functionName, args, to, from) {
        return this.pxe.viewTx(functionName, args, to, from);
    }
    getExtendedContractData(contractAddress) {
        return this.pxe.getExtendedContractData(contractAddress);
    }
    getContractData(contractAddress) {
        return this.pxe.getContractData(contractAddress);
    }
    getUnencryptedLogs(filter) {
        return this.pxe.getUnencryptedLogs(filter);
    }
    getBlockNumber() {
        return this.pxe.getBlockNumber();
    }
    getNodeInfo() {
        return this.pxe.getNodeInfo();
    }
    isGlobalStateSynchronized() {
        return this.pxe.isGlobalStateSynchronized();
    }
    isAccountStateSynchronized(account) {
        return this.pxe.isAccountStateSynchronized(account);
    }
    getSyncStatus() {
        return this.pxe.getSyncStatus();
    }
    addAuthWitness(authWitness) {
        return this.pxe.addAuthWitness(authWitness);
    }
}
