// @aztec/types/dest/notes/extended_note.js
import { AztecAddress, Fr } from '@aztec/circuits.js';
import { BufferReader, Note, TxHash } from '@aztec/types';
/**
 * A note with contextual data.
 */
export class ExtendedNote {
    constructor(
    /** The note as emitted from the Noir contract. */
    note, 
    /** The owner whose public key was used to encrypt the note. */
    owner, 
    /** The contract address this note is created in. */
    contractAddress, 
    /** The specific storage location of the note on the contract. */
    storageSlot, 
    /** The hash of the tx the note was created in. */
    txHash) {
        this.note = note;
        this.owner = owner;
        this.contractAddress = contractAddress;
        this.storageSlot = storageSlot;
        this.txHash = txHash;
    }
    toBuffer() {
        return Buffer.concat([
            this.note.toBuffer(),
            this.owner.toBuffer(),
            this.contractAddress.toBuffer(),
            this.storageSlot.toBuffer(),
            this.txHash.buffer,
        ]);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const note = Note.fromBuffer(reader);
        const owner = AztecAddress.fromBuffer(reader);
        const contractAddress = AztecAddress.fromBuffer(reader);
        const storageSlot = Fr.fromBuffer(reader);
        const txHash = new TxHash(reader.readBytes(TxHash.SIZE));
        return new this(note, owner, contractAddress, storageSlot, txHash);
    }
    toString() {
        return '0x' + this.toBuffer().toString('hex');
    }
    static fromString(str) {
        const hex = str.replace(/^0x/, '');
        return ExtendedNote.fromBuffer(Buffer.from(hex, 'hex'));
    }
}
