// @aztec/aztec.js/dest/account/defaults/entrypoint_payload.js
import { Fr, GeneratorIndex } from '@aztec/circuits.js';
import { padArrayEnd } from '@aztec/foundation/collection';
import { pedersenHash } from '@aztec/foundation/crypto';
import { PackedArguments, emptyFunctionCall } from '@aztec/types';
// These must match the values defined in yarn-project/aztec-nr/aztec/src/entrypoint.nr
export const ACCOUNT_MAX_CALLS = 4;
/** Assembles an entrypoint payload from a set of private and public function calls */
export function buildPayload(calls) {
    const nonce = Fr.random();
    const paddedCalls = padArrayEnd(calls, emptyFunctionCall(), ACCOUNT_MAX_CALLS);
    const packedArguments = [];
    for (const call of paddedCalls) {
        packedArguments.push(PackedArguments.fromArgs(call.args));
    }
    const formattedCalls = paddedCalls.map((call, index) => ({
        // eslint-disable-next-line camelcase
        args_hash: packedArguments[index].hash,
        // eslint-disable-next-line camelcase
        function_selector: call.functionData.selector.toField(),
        // eslint-disable-next-line camelcase
        target_address: call.to.toField(),
        // eslint-disable-next-line camelcase
        is_public: !call.functionData.isPrivate,
    }));
    return {
        payload: {
            // eslint-disable-next-line camelcase
            function_calls: formattedCalls,
            nonce,
        },
        packedArguments,
    };
}
/** Hashes an entrypoint payload to a 32-byte buffer (useful for signing) */
export function hashPayload(payload) {
    return pedersenHash(flattenPayload(payload).map(fr => fr.toBuffer()), GeneratorIndex.SIGNATURE_PAYLOAD);
}
/** Flattens an entrypoint payload */
export function flattenPayload(payload) {
    return [
        ...payload.function_calls.flatMap(call => [
            call.args_hash,
            call.function_selector,
            call.target_address,
            new Fr(call.is_public),
        ]),
        payload.nonce,
    ];
}
