// @aztec/aztec.js/dest/contract/base_contract_interaction.js
import { SentTx } from './sent_tx.js';
/**
 * Base class for an interaction with a contract, be it a deployment, a function call, or a batch.
 * Implements the sequence create/simulate/send.
 */
export class BaseContractInteraction {
    constructor(pxe) {
        this.pxe = pxe;
    }
    /**
     * Simulates a transaction execution request and returns a tx object ready to be sent.
     * @param options - optional arguments to be used in the creation of the transaction
     * @returns The resulting transaction
     */
    async simulate(options = {}) {
        const txRequest = this.txRequest ?? (await this.create(options));
        this.tx = await this.pxe.simulateTx(txRequest, !options.skipPublicSimulation);
        return this.tx;
    }
    /**
     * Sends a transaction to the contract function with the specified options.
     * This function throws an error if called on an unconstrained function.
     * It creates and signs the transaction if necessary, and returns a SentTx instance,
     * which can be used to track the transaction status, receipt, and events.
     * @param options - An optional object containing 'from' property representing
     * the AztecAddress of the sender. If not provided, the default address is used.
     * @returns A SentTx instance for tracking the transaction status and information.
     */
    send(options = {}) {
        const promise = (async () => {
            const tx = this.tx ?? (await this.simulate(options));
            return this.pxe.sendTx(tx);
        })();
        return new SentTx(this.pxe, promise);
    }
}
