// @aztec/aztec.js/dest/contract/contract_base.js
import { FunctionSelector } from '@aztec/foundation/abi';
import { ContractFunctionInteraction } from './contract_function_interaction.js';
/**
 * Abstract implementation of a contract extended by the Contract class and generated contract types.
 */
export class ContractBase {
    constructor(
    /** The deployed contract's complete address. */
    completeAddress, 
    /** The Application Binary Interface for the contract. */
    artifact, 
    /** The wallet used for interacting with this contract. */
    wallet, 
    /** The portal contract address on L1, if any. */
    portalContract) {
        this.completeAddress = completeAddress;
        this.artifact = artifact;
        this.wallet = wallet;
        this.portalContract = portalContract;
        /**
         * An object containing contract methods mapped to their respective names.
         */
        this.methods = {};
        artifact.functions.forEach((f) => {
            const interactionFunction = (...args) => {
                return new ContractFunctionInteraction(this.wallet, this.completeAddress.address, f, args);
            };
            this.methods[f.name] = Object.assign(interactionFunction, {
                /**
                 * A getter for users to fetch the function selector.
                 * @returns Selector of the function.
                 */
                get selector() {
                    return FunctionSelector.fromNameAndParameters(f.name, f.parameters);
                },
            });
        });
    }
    /**
     * Address of the contract.
     */
    get address() {
        return this.completeAddress.address;
    }
    /**
     * Creates a new instance of the contract wrapper attached to a different wallet.
     * @param wallet - Wallet to use for sending txs.
     * @returns A new contract instance.
     */
    withWallet(wallet) {
        return new ContractBase(this.completeAddress, this.artifact, wallet, this.portalContract);
    }
}
