// @aztec/types/dest/auth_witness.js
import { Vector } from '@aztec/circuits.js';
import { BufferReader, serializeToBuffer } from '@aztec/circuits.js/utils';
import { Fr } from '@aztec/foundation/fields';
/**
 * An authentication witness. Used to authorize an action by a user.
 */
export class AuthWitness {
    constructor(
    /** Hash of the request to authorize */
    requestHash, 
    /** Authentication witness for the hash  */
    witness) {
        this.requestHash = requestHash;
        this.witness = witness.map(x => new Fr(x));
    }
    toBuffer() {
        return serializeToBuffer(this.requestHash, new Vector(this.witness));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new AuthWitness(Fr.fromBuffer(reader), reader.readVector(Fr));
    }
    toString() {
        return '0x' + this.toBuffer().toString('hex');
    }
    static fromString(str) {
        const hex = str.replace(/^0x/, '');
        return AuthWitness.fromBuffer(Buffer.from(hex, 'hex'));
    }
}
