// @aztec/types/dest/contract_data.js
import { CompleteAddress, FUNCTION_SELECTOR_NUM_BYTES, Fr, FunctionSelector, Point, } from '@aztec/circuits.js';
import { BufferReader, serializeToBuffer } from '@aztec/circuits.js/utils';
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { randomBytes } from '@aztec/foundation/crypto';
import { EthAddress } from '@aztec/foundation/eth-address';
import { numToInt32BE, serializeBufferArrayToVector } from '@aztec/foundation/serialize';
export { BufferReader } from '@aztec/circuits.js/utils';
/**
 * Represents encoded contract function.
 */
export class EncodedContractFunction {
    constructor(
    /**
     * The function selector.
     */
    selector, 
    /**
     * Whether the function is internal.
     */
    isInternal, 
    /**
     * The function bytecode.
     */
    bytecode) {
        this.selector = selector;
        this.isInternal = isInternal;
        this.bytecode = bytecode;
    }
    /**
     * Serializes this instance into a buffer.
     * @returns Encoded buffer.
     */
    toBuffer() {
        const bytecodeBuf = Buffer.concat([numToInt32BE(this.bytecode.length), this.bytecode]);
        return serializeToBuffer(this.selector, this.isInternal, bytecodeBuf);
    }
    /**
     * Deserializes a contract function object from an encoded buffer.
     * @param buffer - The encoded buffer.
     * @returns The deserialized contract function.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const fnSelector = FunctionSelector.fromBuffer(reader.readBytes(FUNCTION_SELECTOR_NUM_BYTES));
        const isInternal = reader.readBoolean();
        return new EncodedContractFunction(fnSelector, isInternal, reader.readBuffer());
    }
    /**
     * Creates a random contract function.
     * @returns A random contract function.
     */
    static random() {
        return new EncodedContractFunction(FunctionSelector.fromBuffer(randomBytes(4)), false, randomBytes(64));
    }
}
/**
 * A contract data blob, containing L1 and L2 addresses, public functions' bytecode, partial address and public key.
 */
export class ExtendedContractData {
    constructor(
    /** The base contract data: aztec & portal addresses. */
    contractData, 
    /** Artifacts of public functions. */
    publicFunctions, 
    /** Partial addresses of the contract. */
    partialAddress, 
    /** Public key of the contract. */
    publicKey) {
        this.contractData = contractData;
        this.publicFunctions = publicFunctions;
        this.partialAddress = partialAddress;
        this.publicKey = publicKey;
        this.bytecode = serializeBufferArrayToVector(publicFunctions.map(fn => fn.toBuffer()));
    }
    /**
     * Gets the public function data or undefined.
     * @param selector - The function selector of the function to fetch.
     * @returns The public function data (if found).
     */
    getPublicFunction(selector) {
        return this.publicFunctions.find(fn => fn.selector.equals(selector));
    }
    /**
     * Serializes this instance into a buffer, using 20 bytes for the eth address.
     * @returns Encoded buffer.
     */
    toBuffer() {
        const contractDataBuf = this.contractData.toBuffer();
        return serializeToBuffer(contractDataBuf, this.bytecode, this.partialAddress, this.publicKey);
    }
    /**
     * Serializes this instance into a string.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Gets the complete address.
     * @returns The complete address.
     */
    getCompleteAddress() {
        return new CompleteAddress(this.contractData.contractAddress, this.publicKey, this.partialAddress);
    }
    /** True if this represents an empty instance. */
    isEmpty() {
        return (this.contractData.isEmpty() &&
            this.publicFunctions.length === 0 &&
            this.partialAddress.isZero() &&
            this.publicKey.x.isZero() &&
            this.publicKey.y.isZero());
    }
    /**
     * Deserializes a contract data object from an encoded buffer, using 20 bytes for the eth address.
     * @param buffer - Byte array resulting from calling toBuffer.
     * @returns Deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const contractData = reader.readObject(ContractData);
        const publicFns = reader.readVector(EncodedContractFunction);
        const partialAddress = reader.readObject(Fr);
        const publicKey = reader.readObject(Point);
        return new ExtendedContractData(contractData, publicFns, partialAddress, publicKey);
    }
    /**
     * Deserializes a contract data object from an encoded string, using 20 bytes for the eth address.
     * @param str - String resulting from calling toString.
     * @returns Deserialized instance.
     */
    static fromString(str) {
        return ExtendedContractData.fromBuffer(Buffer.from(str, 'hex'));
    }
    /**
     * Generate ContractData with random addresses.
     * @param contractData - Optional contract data to use.
     * @returns A random ExtendedContractData object.
     */
    static random(contractData) {
        return new ExtendedContractData(contractData ?? ContractData.random(), [EncodedContractFunction.random(), EncodedContractFunction.random()], Fr.random(), Point.random());
    }
    /** Generates empty extended contract data. */
    static empty() {
        return new ExtendedContractData(ContractData.empty(), [], Fr.ZERO, Point.ZERO);
    }
}
/**
 * A contract data blob, containing L1 and L2 addresses.
 */
export class ContractData {
    constructor(
    /**
     * The L2 address of the contract, as a field element (32 bytes).
     */
    contractAddress, 
    /**
     * The L1 address of the contract, (20 bytes).
     */
    portalContractAddress) {
        this.contractAddress = contractAddress;
        this.portalContractAddress = portalContractAddress;
    }
    /**
     * Serializes this instance into a buffer, using 20 bytes for the eth address.
     * @returns Encoded buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.contractAddress, this.portalContractAddress.toBuffer20());
    }
    /**
     * Serializes this instance into a string, using 20 bytes for the eth address.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /** True if all data is zero. */
    isEmpty() {
        return this.contractAddress.isZero() && this.portalContractAddress.isZero();
    }
    /**
     * Deserializes a contract data object from an encoded buffer, using 20 bytes for the eth address.
     * @param buffer - Byte array resulting from calling toBuffer.
     * @returns Deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const aztecAddr = AztecAddress.fromBuffer(reader);
        const ethAddr = new EthAddress(reader.readBytes(EthAddress.SIZE_IN_BYTES));
        return new ContractData(aztecAddr, ethAddr);
    }
    /**
     * Deserializes a contract data object from an encoded string, using 20 bytes for the eth address.
     * @param str - String resulting from calling toString.
     * @returns Deserialized instance.
     */
    static fromString(str) {
        return ContractData.fromBuffer(Buffer.from(str, 'hex'));
    }
    /**
     * Generate ContractData with random addresses.
     * @returns ContractData.
     */
    static random() {
        return new ContractData(AztecAddress.random(), EthAddress.random());
    }
    /** Generates an empty ContractData. */
    static empty() {
        return new ContractData(AztecAddress.ZERO, EthAddress.ZERO);
    }
}
