// @aztec/types/dest/l1_to_l2_message.js
import { BufferReader, serializeToBuffer } from '@aztec/circuits.js/utils';
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
/**
 * L1AndL2Message and Index (in the merkle tree) as one type
 */
export class L1ToL2MessageAndIndex {
    constructor(
    /** the index in the L1 to L2 Message tree. */
    index, 
    /** The message. */
    message) {
        this.index = index;
        this.message = message;
    }
    toBuffer() {
        return Buffer.concat([toBufferBE(this.index, 32), this.message.toBuffer()]);
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(data) {
        const buffer = Buffer.from(data, 'hex');
        return L1ToL2MessageAndIndex.fromBuffer(buffer);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const index = toBigIntBE(reader.readBytes(32));
        const message = L1ToL2Message.fromBuffer(reader);
        return new L1ToL2MessageAndIndex(index, message);
    }
}
/**
 * An L1 to L2 message emitted in a particular L1 block.
 */
export class PendingL1ToL2Message {
    constructor(
    /** the message */
    message, 
    /** the L1 block this message was emitted in */
    blockNumber, 
    /** at which index in the L1 block this message was emitted */
    indexInBlock) {
        this.message = message;
        this.blockNumber = blockNumber;
        this.indexInBlock = indexInBlock;
    }
}
/**
 * An L1 to L2 message that was cancelled.
 */
export class CancelledL1ToL2Message {
    constructor(
    /** the message */
    entryKey, 
    /** the L1 block this message was emitted in */
    blockNumber, 
    /** at which index in the L1 block this message was emitted */
    indexInBlock) {
        this.entryKey = entryKey;
        this.blockNumber = blockNumber;
        this.indexInBlock = indexInBlock;
    }
}
/**
 * The format of an L1 to L2 Message.
 */
export class L1ToL2Message {
    constructor(
    /**
     * The sender of the message on L1.
     */
    sender, 
    /**
     * The recipient of the message on L2.
     */
    recipient, 
    /**
     * The message content.
     */
    content, 
    /**
     * The hash of the spending secret.
     */
    secretHash, 
    /**
     * The deadline for the message.
     */
    deadline, 
    /**
     * The fee for the message.
     */
    fee, 
    /**
     * The entry key for the message - optional.
     */
    entryKey) {
        this.sender = sender;
        this.recipient = recipient;
        this.content = content;
        this.secretHash = secretHash;
        this.deadline = deadline;
        this.fee = fee;
        this.entryKey = entryKey;
    }
    /**
     * Returns each element within its own field so that it can be consumed by an acvm oracle call.
     * @returns The message as an array of fields (in order).
     */
    toFieldArray() {
        return [
            ...this.sender.toFieldArray(),
            ...this.recipient.toFieldArray(),
            this.content,
            this.secretHash,
            new Fr(BigInt(this.deadline)),
            new Fr(BigInt(this.fee)),
        ];
    }
    toBuffer() {
        return serializeToBuffer(this.sender, this.recipient, this.content, this.secretHash, this.deadline, this.fee);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const sender = reader.readObject(L1Actor);
        const recipient = reader.readObject(L2Actor);
        const content = Fr.fromBuffer(reader);
        const secretHash = Fr.fromBuffer(reader);
        const deadline = reader.readNumber();
        const fee = reader.readNumber();
        return new L1ToL2Message(sender, recipient, content, secretHash, deadline, fee);
    }
    static empty() {
        return new L1ToL2Message(L1Actor.empty(), L2Actor.empty(), Fr.ZERO, Fr.ZERO, 0, 0);
    }
    static random(entryKey) {
        return new L1ToL2Message(L1Actor.random(), L2Actor.random(), Fr.random(), Fr.random(), Math.floor(Math.random() * 1000), Math.floor(Math.random() * 1000), entryKey);
    }
}
/**
 * The sender of an L1 to L2 message.
 */
export class L1Actor {
    constructor(
    /**
     * The sender of the message.
     */
    sender, 
    /**
     * The chain id on which the message was sent.
     */
    chainId) {
        this.sender = sender;
        this.chainId = chainId;
    }
    static empty() {
        return new L1Actor(EthAddress.ZERO, 0);
    }
    toFieldArray() {
        return [this.sender.toField(), new Fr(BigInt(this.chainId))];
    }
    toBuffer() {
        return serializeToBuffer(this.sender, this.chainId);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const ethAddr = new EthAddress(reader.readBytes(32));
        const chainId = reader.readNumber();
        return new L1Actor(ethAddr, chainId);
    }
    static random() {
        return new L1Actor(EthAddress.random(), Math.floor(Math.random() * 1000));
    }
}
/**
 * The recipient of an L2 message.
 */
export class L2Actor {
    constructor(
    /**
     * The recipient of the message.
     */
    recipient, 
    /**
     * The version of the protocol.
     */
    version) {
        this.recipient = recipient;
        this.version = version;
    }
    static empty() {
        return new L2Actor(AztecAddress.ZERO, 0);
    }
    toFieldArray() {
        return [this.recipient.toField(), new Fr(BigInt(this.version))];
    }
    toBuffer() {
        return serializeToBuffer(this.recipient, this.version);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const aztecAddr = AztecAddress.fromBuffer(reader);
        const version = reader.readNumber();
        return new L2Actor(aztecAddr, version);
    }
    static random() {
        return new L2Actor(AztecAddress.random(), Math.floor(Math.random() * 1000));
    }
}
