// @aztec/types/dest/packed_arguments.js
import { Fr, Vector } from '@aztec/circuits.js';
import { computeVarArgsHash } from '@aztec/circuits.js/abis';
import { BufferReader, serializeToBuffer } from '@aztec/circuits.js/utils';
/**
 * Packs a set of arguments into a hash.
 */
export class PackedArguments {
    constructor(
    /**
     *  Function arguments.
     */
    args, 
    /**
     * The hash of the args
     */
    hash) {
        this.args = args;
        this.hash = hash;
    }
    static getFields(fields) {
        return [fields.args, fields.hash];
    }
    static from(fields) {
        return new PackedArguments(...PackedArguments.getFields(fields));
    }
    static fromArgs(args) {
        return new PackedArguments(args, computeVarArgsHash(args));
    }
    toBuffer() {
        return serializeToBuffer(new Vector(this.args), this.hash);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PackedArguments(reader.readVector(Fr), Fr.fromBuffer(reader));
    }
}
