// @aztec/types/dest/tx_execution_request.js
import { AztecAddress, Fr, FunctionData, TxContext, TxRequest, Vector } from '@aztec/circuits.js';
import { BufferReader, serializeToBuffer } from '@aztec/circuits.js/utils';
import { AuthWitness } from './auth_witness.js';
import { PackedArguments } from './packed_arguments.js';
/**
 * Request to execute a transaction. Similar to TxRequest, but has the full args.
 */
export class TxExecutionRequest {
    constructor(
    /**
     * Sender.
     */
    origin, 
    /**
     * Function data representing the function to call.
     * TODO(#3417): Remove this field and replace with a function selector.
     */
    functionData, 
    /**
     * The hash of the entry point arguments.
     */
    argsHash, 
    /**
     * Transaction context.
     */
    txContext, 
    /**
     * These packed arguments will be used during transaction simulation.
     * For example, a call to an account contract might contain as many packed arguments
     * as relayed function calls, and one for the entrypoint.
     */
    packedArguments, 
    /**
     * Transient authorization witnesses for authorizing the execution of one or more actions during this tx.
     * These witnesses are not expected to be stored in the local witnesses database of the PXE.
     */
    authWitnesses) {
        this.origin = origin;
        this.functionData = functionData;
        this.argsHash = argsHash;
        this.txContext = txContext;
        this.packedArguments = packedArguments;
        this.authWitnesses = authWitnesses;
    }
    toTxRequest() {
        return new TxRequest(this.origin, this.functionData, this.argsHash, this.txContext);
    }
    static getFields(fields) {
        return [
            fields.origin,
            fields.functionData,
            fields.argsHash,
            fields.txContext,
            fields.packedArguments,
            fields.authWitnesses,
        ];
    }
    static from(fields) {
        return new TxExecutionRequest(...TxExecutionRequest.getFields(fields));
    }
    /**
     * Serialize as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.origin, this.functionData, this.argsHash, this.txContext, new Vector(this.packedArguments), new Vector(this.authWitnesses));
    }
    /**
     * Serialize as a string.
     * @returns The string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer to read from.
     * @returns The deserialized TxRequest object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new TxExecutionRequest(reader.readObject(AztecAddress), reader.readObject(FunctionData), Fr.fromBuffer(reader), reader.readObject(TxContext), reader.readVector(PackedArguments), reader.readVector(AuthWitness));
    }
    /**
     * Deserializes from a string, corresponding to a write in cpp.
     * @param str - String to read from.
     * @returns The deserialized TxRequest object.
     */
    static fromString(str) {
        return TxExecutionRequest.fromBuffer(Buffer.from(str, 'hex'));
    }
}
