// @aztec/foundation/dest/running-promise/index.js
/**
 * RunningPromise is a utility class that helps manage the execution of an asynchronous function
 * at a specified polling interval. It allows starting, stopping, and checking the status of the
 * internally managed promise. The class also supports interrupting the polling process when stopped.
 */
export class RunningPromise {
    constructor(fn, pollingInterval = 10000) {
        this.fn = fn;
        this.pollingInterval = pollingInterval;
        this.running = false;
        this.runningPromise = Promise.resolve();
        this.interruptPromise = Promise.resolve();
        this.interruptResolve = () => { };
    }
    /**
     * Starts the running promise.
     */
    start() {
        this.running = true;
        this.interruptPromise = new Promise(resolve => (this.interruptResolve = resolve));
        const poll = async () => {
            while (this.running) {
                await this.fn();
                await this.interruptibleSleep(this.pollingInterval);
            }
        };
        this.runningPromise = poll();
    }
    /**
     * Stops the running promise, resolves any pending interruptible sleep,
     * and waits for the currently executing function to complete.
     */
    async stop() {
        this.running = false;
        this.interruptResolve();
        await this.runningPromise;
    }
    /**
     * A sleep function that can be interrupted before the specified time.
     * The sleep duration is determined by 'timeInMs', and it can be terminated early if the 'interruptPromise' is resolved.
     * @param timeInMs - The time in milliseconds.
     */
    async interruptibleSleep(timeInMs) {
        let timeout;
        const sleepPromise = new Promise(resolve => {
            timeout = setTimeout(resolve, timeInMs);
        });
        await Promise.race([sleepPromise, this.interruptPromise]);
        clearTimeout(timeout);
    }
    /**
     * Checks if the running promise is currently active.
     * @returns True if the promise is running.
     */
    isRunning() {
        return this.running;
    }
}
