// @aztec/foundation/dest/abi/abi.js
import { inflate } from 'pako';
/**
 * Indicates whether a parameter is public or secret/private.
 */
export var ABIParameterVisibility;
(function (ABIParameterVisibility) {
    ABIParameterVisibility["PUBLIC"] = "public";
    ABIParameterVisibility["SECRET"] = "secret";
})(ABIParameterVisibility || (ABIParameterVisibility = {}));
/**
 * Aztec.nr function types.
 */
export var FunctionType;
(function (FunctionType) {
    FunctionType["SECRET"] = "secret";
    FunctionType["OPEN"] = "open";
    FunctionType["UNCONSTRAINED"] = "unconstrained";
})(FunctionType || (FunctionType = {}));
/**
 * Gets the debug metadata of a given function from the contract artifact
 * @param artifact - The contract build artifact
 * @param functionName - The name of the function
 * @returns The debug metadata of the function
 */
export function getFunctionDebugMetadata(artifact, functionName) {
    const functionIndex = artifact.functions.findIndex(f => f.name === functionName);
    if (artifact.debug && functionIndex !== -1) {
        const debugSymbols = JSON.parse(inflate(Buffer.from(artifact.debug.debugSymbols[functionIndex], 'base64'), { to: 'string' }));
        const files = artifact.debug.fileMap;
        return {
            debugSymbols,
            files,
        };
    }
    return undefined;
}
