// @aztec/circuits.js/dest/structs/kernel/private_kernel.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { CONTRACT_TREE_HEIGHT, FUNCTION_TREE_HEIGHT, MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL, MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, MAX_READ_REQUESTS_PER_CALL, } from '../../constants.gen.js';
import { serializeToBuffer } from '../../utils/serialize.js';
import { CallRequest } from '../call_request.js';
import { PrivateCallStackItem } from '../call_stack_item.js';
import { MembershipWitness } from '../membership_witness.js';
import { Proof } from '../proof.js';
import { ReadRequestMembershipWitness } from '../read_request_membership_witness.js';
import { VerificationKey } from '../verification_key.js';
import { PreviousKernelData } from './previous_kernel_data.js';
/**
 * Private call data.
 */
export class PrivateCallData {
    constructor(
    /**
     * The call stack item currently being processed.
     */
    callStackItem, 
    /**
     * Other private call stack items to be processed.
     */
    privateCallStack, 
    /**
     * Other public call stack items to be processed.
     */
    publicCallStack, 
    /**
     * The proof of the execution of this private call.
     */
    proof, 
    /**
     * The verification key for the function being invoked.
     */
    vk, 
    /**
     * The membership witness for the function leaf corresponding to the function being invoked.
     */
    functionLeafMembershipWitness, 
    /**
     * The membership witness for the contract leaf corresponding to the contract on which the function is being
     * invoked.
     */
    contractLeafMembershipWitness, 
    /**
     * The membership witnesses for read requests created by the function being invoked.
     */
    readRequestMembershipWitnesses, 
    /**
     * The address of the portal contract corresponding to the contract on which the function is being invoked.
     */
    portalContractAddress, 
    /**
     * The hash of the ACIR of the function being invoked.
     */
    acirHash) {
        this.callStackItem = callStackItem;
        this.privateCallStack = privateCallStack;
        this.publicCallStack = publicCallStack;
        this.proof = proof;
        this.vk = vk;
        this.functionLeafMembershipWitness = functionLeafMembershipWitness;
        this.contractLeafMembershipWitness = contractLeafMembershipWitness;
        this.readRequestMembershipWitnesses = readRequestMembershipWitnesses;
        this.portalContractAddress = portalContractAddress;
        this.acirHash = acirHash;
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            fields.callStackItem,
            fields.privateCallStack,
            fields.publicCallStack,
            fields.proof,
            fields.vk,
            fields.functionLeafMembershipWitness,
            fields.contractLeafMembershipWitness,
            fields.readRequestMembershipWitnesses,
            fields.portalContractAddress,
            fields.acirHash,
        ];
    }
    static from(fields) {
        return new PrivateCallData(...PrivateCallData.getFields(fields));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...PrivateCallData.getFields(this));
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateCallData(reader.readObject(PrivateCallStackItem), reader.readArray(MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL, CallRequest), reader.readArray(MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, CallRequest), reader.readObject(Proof), reader.readObject(VerificationKey), reader.readObject(MembershipWitness.deserializer(FUNCTION_TREE_HEIGHT)), reader.readObject(MembershipWitness.deserializer(CONTRACT_TREE_HEIGHT)), reader.readArray(MAX_READ_REQUESTS_PER_CALL, ReadRequestMembershipWitness), reader.readObject(Fr), reader.readObject(Fr));
    }
}
/**
 * Input to the private kernel circuit - initial call.
 */
export class PrivateKernelInputsInit {
    constructor(
    /**
     * The transaction request which led to the creation of these inputs.
     */
    txRequest, 
    /**
     * Private calldata corresponding to this iteration of the kernel.
     */
    privateCall) {
        this.txRequest = txRequest;
        this.privateCall = privateCall;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.txRequest, this.privateCall);
    }
}
/**
 * Input to the private kernel circuit - Inner call.
 */
export class PrivateKernelInputsInner {
    constructor(
    /**
     * The previous kernel data (dummy if this is the first kernel).
     */
    previousKernel, 
    /**
     * Private calldata corresponding to this iteration of the kernel.
     */
    privateCall) {
        this.previousKernel = previousKernel;
        this.privateCall = privateCall;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.privateCall);
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateKernelInputsInner(reader.readObject(PreviousKernelData), reader.readObject(PrivateCallData));
    }
}
/**
 * Input to the private kernel circuit - Final ordering call.
 */
export class PrivateKernelInputsOrdering {
    constructor(
    /**
     * The previous kernel data
     */
    previousKernel, 
    /**
     * Contains hints for the transient read requests to localize corresponding commitments.
     */
    readCommitmentHints, 
    /**
     * Contains hints for the transient nullifiers to localize corresponding commitments.
     */
    nullifierCommitmentHints) {
        this.previousKernel = previousKernel;
        this.readCommitmentHints = readCommitmentHints;
        this.nullifierCommitmentHints = nullifierCommitmentHints;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.readCommitmentHints);
    }
}
