// @aztec/circuits.js/dest/structs/kernel/public_inputs_final.js
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../../utils/serialize.js';
import { FinalAccumulatedData } from './combined_accumulated_data.js';
import { CombinedConstantData } from './combined_constant_data.js';
/**
 * Public inputs of the final ordering private kernel circuit.
 */
export class KernelCircuitPublicInputsFinal {
    constructor(
    /**
     * Final data accumulated for ordering private kernel circuit.
     */
    end, 
    /**
     * Data which is not modified by the circuits.
     */
    constants, 
    /**
     * Indicates whether the input is for a private or public kernel.
     */
    isPrivate) {
        this.end = end;
        this.constants = constants;
        this.isPrivate = isPrivate;
    }
    toBuffer() {
        return serializeToBuffer(this.end, this.constants, this.isPrivate);
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of KernelCircuitPublicInputsFinal.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new KernelCircuitPublicInputsFinal(reader.readObject(FinalAccumulatedData), reader.readObject(CombinedConstantData), reader.readBoolean());
    }
    static empty() {
        return new KernelCircuitPublicInputsFinal(FinalAccumulatedData.empty(), CombinedConstantData.empty(), true);
    }
}
/**
 * Public inputs of the final private kernel circuit.
 */
export class PrivateKernelPublicInputsFinal extends KernelCircuitPublicInputsFinal {
    constructor(end, constants) {
        super(end, constants, true);
    }
}
