// @aztec/types/dest/tx/tx_hash.js
import { deserializeBigInt, serializeBigInt } from '@aztec/foundation/serialize';
/**
 * A class representing hash of Aztec transaction.
 */
export class TxHash {
    constructor(
    /**
     * The buffer containing the hash.
     */
    buffer) {
        this.buffer = buffer;
        if (buffer.length !== TxHash.SIZE) {
            throw new Error(`Expected buffer to have length ${TxHash.SIZE} but was ${buffer.length}`);
        }
    }
    /**
     * Checks if this hash and another hash are equal.
     * @param hash - A hash to compare with.
     * @returns True if the hashes are equal, false otherwise.
     */
    equals(hash) {
        return this.buffer.equals(hash.buffer);
    }
    /**
     * Returns true if this hash is zero.
     * @returns True if this hash is zero.
     */
    isZero() {
        return this.buffer.equals(Buffer.alloc(32, 0));
    }
    /**
     * Convert this hash to a hex string.
     * @returns The hex string.
     */
    toString() {
        return this.buffer.toString('hex');
    }
    /**
     * Convert this hash to a big int.
     * @returns The big int.
     */
    toBigInt() {
        return deserializeBigInt(this.buffer, 0, TxHash.SIZE).elem;
    }
    /**
     * Creates a tx hash from a bigint.
     * @param hash - The tx hash as a big int.
     * @returns The TxHash.
     */
    static fromBigInt(hash) {
        return new TxHash(serializeBigInt(hash, TxHash.SIZE));
    }
    /**
     * Converts this hash from a buffer of 28 bytes.
     * Verifies the input is 28 bytes.
     * @param buffer - The 28 byte buffer to construct from.
     * @returns A TxHash created from the input buffer with 4 bytes 0 padding at the front.
     */
    static fromBuffer28(buffer) {
        if (buffer.length != 28) {
            throw new Error(`Expected TxHash input buffer to be 28 bytes`);
        }
        const padded = Buffer.concat([Buffer.alloc(this.SIZE - 28), buffer]);
        return new TxHash(padded);
    }
    /**
     * Converts a string into a TxHash object.
     * @param str - The TX hash in string format.
     * @returns A new TxHash object.
     */
    static fromString(str) {
        return new TxHash(Buffer.from(str, 'hex'));
    }
}
/**
 * The size of the hash in bytes.
 */
TxHash.SIZE = 32;
/**
 * TxHash with value zero.
 */
TxHash.ZERO = new TxHash(Buffer.alloc(TxHash.SIZE));
