// @aztec/types/dest/tx/tx_receipt.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { TxHash } from '@aztec/types';
/**
 * Possible status of a transaction.
 */
export var TxStatus;
(function (TxStatus) {
    TxStatus["DROPPED"] = "dropped";
    TxStatus["MINED"] = "mined";
    TxStatus["PENDING"] = "pending";
})(TxStatus || (TxStatus = {}));
/**
 * Represents a transaction receipt in the Aztec network.
 * Contains essential information about the transaction including its status, origin, and associated addresses.
 */
export class TxReceipt {
    constructor(
    /**
     * A unique identifier for a transaction.
     */
    txHash, 
    /**
     * The transaction's status.
     */
    status, 
    /**
     * Description of transaction error, if any.
     */
    error, 
    /**
     * The hash of the block containing the transaction.
     */
    blockHash, 
    /**
     * The block number in which the transaction was included.
     */
    blockNumber, 
    /**
     * The deployed contract's address.
     */
    contractAddress, 
    /**
     * Information useful for testing/debugging, set when test flag is set to true in `waitOpts`.
     */
    debugInfo) {
        this.txHash = txHash;
        this.status = status;
        this.error = error;
        this.blockHash = blockHash;
        this.blockNumber = blockNumber;
        this.contractAddress = contractAddress;
        this.debugInfo = debugInfo;
    }
    /**
     * Convert a Tx class object to a plain JSON object.
     * @returns A plain object with Tx properties.
     */
    toJSON() {
        return {
            txHash: this.txHash.toString(),
            status: this.status.toString(),
            error: this.error,
            blockHash: this.blockHash?.toString('hex'),
            blockNumber: this.blockNumber,
            contractAddress: this.contractAddress?.toString(),
        };
    }
    /**
     * Convert a plain JSON object to a Tx class object.
     * @param obj - A plain Tx JSON object.
     * @returns A Tx class object.
     */
    static fromJSON(obj) {
        const txHash = TxHash.fromString(obj.txHash);
        const status = obj.status;
        const error = obj.error;
        const blockHash = obj.blockHash ? Buffer.from(obj.blockHash, 'hex') : undefined;
        const blockNumber = obj.blockNumber ? Number(obj.blockNumber) : undefined;
        const contractAddress = obj.contractAddress ? AztecAddress.fromString(obj.contractAddress) : undefined;
        return new TxReceipt(txHash, status, error, blockHash, blockNumber, contractAddress);
    }
}
