// @aztec/circuits.js/dest/structs/rollup/base_rollup.js
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../../utils/serialize.js';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
/**
 * Class containing the data of a preimage of a single leaf in the nullifier tree.
 * Note: It's called preimage because this data gets hashed before being inserted as a node into the `IndexedTree`.
 */
export class NullifierLeafPreimage {
    constructor(
    /**
     * Leaf value inside the indexed tree's linked list.
     */
    nullifier, 
    /**
     * Next value inside the indexed tree's linked list.
     */
    nextNullifier, 
    /**
     * Index of the next leaf in the indexed tree's linked list.
     */
    nextIndex) {
        this.nullifier = nullifier;
        this.nextNullifier = nextNullifier;
        this.nextIndex = nextIndex;
    }
    getKey() {
        return this.nullifier.toBigInt();
    }
    getNextKey() {
        return this.nextNullifier.toBigInt();
    }
    getNextIndex() {
        return this.nextIndex;
    }
    asLeaf() {
        return new NullifierLeaf(this.nullifier);
    }
    toBuffer() {
        return Buffer.concat(this.toHashInputs());
    }
    toHashInputs() {
        return [
            Buffer.from(this.nullifier.toBuffer()),
            Buffer.from(toBufferBE(this.nextIndex, 32)),
            Buffer.from(this.nextNullifier.toBuffer()),
        ];
    }
    clone() {
        return new NullifierLeafPreimage(this.nullifier, this.nextNullifier, this.nextIndex);
    }
    static empty() {
        return new NullifierLeafPreimage(Fr.ZERO, Fr.ZERO, 0n);
    }
    static fromBuffer(buf) {
        const nullifier = Fr.fromBuffer(buf.subarray(0, 32));
        const nextIndex = toBigIntBE(buf.subarray(32, 64));
        const nextNullifier = Fr.fromBuffer(buf.subarray(64, 96));
        return new NullifierLeafPreimage(nullifier, nextNullifier, nextIndex);
    }
    static fromLeaf(leaf, nextKey, nextIndex) {
        return new NullifierLeafPreimage(leaf.nullifier, new Fr(nextKey), nextIndex);
    }
    static clone(preimage) {
        return new NullifierLeafPreimage(preimage.nullifier, preimage.nextNullifier, preimage.nextIndex);
    }
}
/**
 * A nullifier to be inserted in the nullifier tree.
 */
export class NullifierLeaf {
    constructor(
    /**
     * Nullifier value.
     */
    nullifier) {
        this.nullifier = nullifier;
    }
    getKey() {
        return this.nullifier.toBigInt();
    }
    toBuffer() {
        return this.nullifier.toBuffer();
    }
    isEmpty() {
        return this.nullifier.isZero();
    }
    static buildDummy(key) {
        return new NullifierLeaf(new Fr(key));
    }
    static fromBuffer(buf) {
        return new NullifierLeaf(Fr.fromBuffer(buf));
    }
}
/**
 * Data which is forwarded through the base rollup circuits unchanged.
 */
export class ConstantRollupData {
    constructor(
    /**
     * Snapshot of the blocks tree at the start of the rollup.
     */
    archiveSnapshot, 
    /**
     * Root of the private kernel verification key tree.
     */
    privateKernelVkTreeRoot, 
    /**
     * Root of the public kernel circuit verification key tree.
     */
    publicKernelVkTreeRoot, 
    /**
     * Hash of the base rollup circuit verification key.
     */
    baseRollupVkHash, 
    /**
     * Hash of the merge rollup circuit verification key.
     */
    mergeRollupVkHash, 
    /**
     * Global variables for the block
     */
    globalVariables) {
        this.archiveSnapshot = archiveSnapshot;
        this.privateKernelVkTreeRoot = privateKernelVkTreeRoot;
        this.publicKernelVkTreeRoot = publicKernelVkTreeRoot;
        this.baseRollupVkHash = baseRollupVkHash;
        this.mergeRollupVkHash = mergeRollupVkHash;
        this.globalVariables = globalVariables;
    }
    static from(fields) {
        return new ConstantRollupData(...ConstantRollupData.getFields(fields));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ConstantRollupData(reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(GlobalVariables));
    }
    static getFields(fields) {
        return [
            fields.archiveSnapshot,
            fields.privateKernelVkTreeRoot,
            fields.publicKernelVkTreeRoot,
            fields.baseRollupVkHash,
            fields.mergeRollupVkHash,
            fields.globalVariables,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...ConstantRollupData.getFields(this));
    }
}
/**
 * Inputs to the base rollup circuit.
 */
export class BaseRollupInputs {
    constructor(
    /**
     * Data of the 2 kernels that preceded this base rollup circuit.
     */
    kernelData, 
    /**
     * Snapshot of the note hash tree at the start of the base rollup circuit.
     */
    startNoteHashTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the start of the base rollup circuit.
     */
    startNullifierTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the start of the base rollup circuit.
     */
    startContractTreeSnapshot, 
    /**
     * Root of the public data tree at the start of the base rollup circuit.
     */
    startPublicDataTreeRoot, 
    /**
     * Snapshot of the blocks tree at the start of the base rollup circuit.
     */
    archiveSnapshot, 
    /**
     * The nullifiers to be inserted in the tree, sorted high to low.
     */
    sortedNewNullifiers, 
    /**
     * The indexes of the sorted nullifiers to the original ones.
     */
    sortednewNullifiersIndexes, 
    /**
     * The nullifiers which need to be updated to perform the batch insertion of the new nullifiers.
     * See `StandardIndexedTree.batchInsert` function for more details.
     */
    lowNullifierLeafPreimages, 
    /**
     * Membership witnesses for the nullifiers which need to be updated to perform the batch insertion of the new
     * nullifiers.
     */
    lowNullifierMembershipWitness, 
    /**
     * Sibling path "pointing to" where the new commitments subtree should be inserted into the note hash tree.
     */
    newCommitmentsSubtreeSiblingPath, 
    /**
     * Sibling path "pointing to" where the new nullifiers subtree should be inserted into the nullifier tree.
     */
    newNullifiersSubtreeSiblingPath, 
    /**
     * Sibling path "pointing to" where the new contracts subtree should be inserted into the contract tree.
     */
    newContractsSubtreeSiblingPath, 
    /**
     * Sibling paths of leaves which are to be affected by the public data update requests.
     * Each item in the array is the sibling path that corresponds to an update request.
     */
    newPublicDataUpdateRequestsSiblingPaths, 
    /**
     * Sibling paths of leaves which are to be read by the public data reads.
     * Each item in the array is the sibling path that corresponds to a read request.
     */
    newPublicDataReadsSiblingPaths, 
    /**
     * Membership witnesses of blocks referred by each of the 2 kernels.
     */
    archiveRootMembershipWitnesses, 
    /**
     * Data which is not modified by the base rollup circuit.
     */
    constants) {
        this.kernelData = kernelData;
        this.startNoteHashTreeSnapshot = startNoteHashTreeSnapshot;
        this.startNullifierTreeSnapshot = startNullifierTreeSnapshot;
        this.startContractTreeSnapshot = startContractTreeSnapshot;
        this.startPublicDataTreeRoot = startPublicDataTreeRoot;
        this.archiveSnapshot = archiveSnapshot;
        this.sortedNewNullifiers = sortedNewNullifiers;
        this.sortednewNullifiersIndexes = sortednewNullifiersIndexes;
        this.lowNullifierLeafPreimages = lowNullifierLeafPreimages;
        this.lowNullifierMembershipWitness = lowNullifierMembershipWitness;
        this.newCommitmentsSubtreeSiblingPath = newCommitmentsSubtreeSiblingPath;
        this.newNullifiersSubtreeSiblingPath = newNullifiersSubtreeSiblingPath;
        this.newContractsSubtreeSiblingPath = newContractsSubtreeSiblingPath;
        this.newPublicDataUpdateRequestsSiblingPaths = newPublicDataUpdateRequestsSiblingPaths;
        this.newPublicDataReadsSiblingPaths = newPublicDataReadsSiblingPaths;
        this.archiveRootMembershipWitnesses = archiveRootMembershipWitnesses;
        this.constants = constants;
    }
    static from(fields) {
        return new BaseRollupInputs(...BaseRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.kernelData,
            fields.startNoteHashTreeSnapshot,
            fields.startNullifierTreeSnapshot,
            fields.startContractTreeSnapshot,
            fields.startPublicDataTreeRoot,
            fields.archiveSnapshot,
            fields.sortedNewNullifiers,
            fields.sortednewNullifiersIndexes,
            fields.lowNullifierLeafPreimages,
            fields.lowNullifierMembershipWitness,
            fields.newCommitmentsSubtreeSiblingPath,
            fields.newNullifiersSubtreeSiblingPath,
            fields.newContractsSubtreeSiblingPath,
            fields.newPublicDataUpdateRequestsSiblingPaths,
            fields.newPublicDataReadsSiblingPaths,
            fields.archiveRootMembershipWitnesses,
            fields.constants,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...BaseRollupInputs.getFields(this));
    }
}
