// @aztec/circuits.js/dest/structs/rollup/base_or_merge_rollup_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { NUM_FIELDS_PER_SHA256 } from '../../constants.gen.js';
import { serializeToBuffer } from '../../utils/serialize.js';
import { AggregationObject } from '../aggregation_object.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
import { ConstantRollupData } from './base_rollup.js';
/**
 * Output of the base and merge rollup circuits.
 */
export class BaseOrMergeRollupPublicInputs {
    constructor(
    /**
     * Specifies from which type of rollup circuit these inputs are from.
     */
    rollupType, 
    /**
     * Rollup sub tree height.
     * Note 1: Base rollup circuit always have a sub tree height of 0.
     * Note 2: With each merge, the sub tree height increases by 1.
     */
    rollupSubtreeHeight, 
    /**
     * Native aggregation state at the end of the rollup circuit.
     */
    endAggregationObject, 
    /**
     * Data which is forwarded through the rollup circuits unchanged.
     */
    constants, 
    /**
     * Snapshot of the note hash tree at the start of the rollup circuit.
     */
    startNoteHashTreeSnapshot, 
    /**
     * Snapshot of the note hash tree at the end of the rollup circuit.
     */
    endNoteHashTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the start of the rollup circuit.
     */
    startNullifierTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the end of the rollup circuit.
     */
    endNullifierTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the start of the rollup circuit.
     */
    startContractTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the end of the rollup circuit.
     */
    endContractTreeSnapshot, 
    /**
     * Root of the public data tree at the start of the rollup circuit.
     */
    startPublicDataTreeRoot, 
    /**
     * Root of the public data tree at the end of the rollup circuit.
     */
    endPublicDataTreeRoot, 
    /**
     * SHA256 hashes of calldata. Used to make public inputs constant-sized (to then be unpacked on-chain).
     * Note: Length 2 for high and low.
     */
    calldataHash) {
        this.rollupType = rollupType;
        this.rollupSubtreeHeight = rollupSubtreeHeight;
        this.endAggregationObject = endAggregationObject;
        this.constants = constants;
        this.startNoteHashTreeSnapshot = startNoteHashTreeSnapshot;
        this.endNoteHashTreeSnapshot = endNoteHashTreeSnapshot;
        this.startNullifierTreeSnapshot = startNullifierTreeSnapshot;
        this.endNullifierTreeSnapshot = endNullifierTreeSnapshot;
        this.startContractTreeSnapshot = startContractTreeSnapshot;
        this.endContractTreeSnapshot = endContractTreeSnapshot;
        this.startPublicDataTreeRoot = startPublicDataTreeRoot;
        this.endPublicDataTreeRoot = endPublicDataTreeRoot;
        this.calldataHash = calldataHash;
    }
    /**
     * Deserializes from a buffer or reader.
     * Note: Corresponds to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized public inputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BaseOrMergeRollupPublicInputs(reader.readNumber(), Fr.fromBuffer(reader), reader.readObject(AggregationObject), reader.readObject(ConstantRollupData), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readArray(NUM_FIELDS_PER_SHA256, Fr));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.rollupType, this.rollupSubtreeHeight, this.endAggregationObject, this.constants, this.startNoteHashTreeSnapshot, this.endNoteHashTreeSnapshot, this.startNullifierTreeSnapshot, this.endNullifierTreeSnapshot, this.startContractTreeSnapshot, this.endContractTreeSnapshot, this.startPublicDataTreeRoot, this.endPublicDataTreeRoot, this.calldataHash);
    }
}
