// @aztec/circuits.js/dest/structs/rollup/previous_rollup_data.js
import { serializeToBuffer } from '../../utils/serialize.js';
/**
 * Represents the data of a previous merge or base rollup circuit.
 */
export class PreviousRollupData {
    constructor(
    /**
     * Public inputs to the base or merge rollup circuit.
     */
    baseOrMergeRollupPublicInputs, 
    /**
     * The proof of the base or merge rollup circuit.
     */
    proof, 
    /**
     * The verification key of the base or merge rollup circuit.
     */
    vk, 
    /**
     * The index of the rollup circuit's vk in a big tree of rollup circuit vks.
     */
    vkIndex, 
    /**
     * Sibling path of the rollup circuit's vk in a big tree of rollup circuit vks.
     */
    vkSiblingPath) {
        this.baseOrMergeRollupPublicInputs = baseOrMergeRollupPublicInputs;
        this.proof = proof;
        this.vk = vk;
        this.vkIndex = vkIndex;
        this.vkSiblingPath = vkSiblingPath;
    }
    /**
     * Serializes previous rollup data to a buffer.
     * @returns The buffer of the serialized previous rollup data.
     */
    toBuffer() {
        return serializeToBuffer(this.baseOrMergeRollupPublicInputs, this.proof, this.vk, this.vkIndex, this.vkSiblingPath);
    }
}
