// @aztec/circuits.js/dest/structs/rollup/root_rollup.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../../utils/serialize.js';
import { AggregationObject } from '../aggregation_object.js';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
/**
 * Represents inputs of the root rollup circuit.
 */
export class RootRollupInputs {
    constructor(
    /**
     * The previous rollup data.
     * Note: Root rollup circuit is the latest circuit the chain of circuits and the previous rollup data is the data
     * from 2 merge or base rollup circuits.
     */
    previousRollupData, 
    /**
     * New L1 to L2 messages.
     */
    newL1ToL2Messages, 
    /**
     * Sibling path of the new L1 to L2 message tree root.
     */
    newL1ToL2MessagesTreeRootSiblingPath, 
    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    startL1ToL2MessagesTreeSnapshot, 
    /**
     * Snapshot of the historical block roots tree at the start of the rollup.
     */
    startArchiveSnapshot, 
    /**
     * Sibling path of the new block tree root.
     */
    newArchiveSiblingPath) {
        this.previousRollupData = previousRollupData;
        this.newL1ToL2Messages = newL1ToL2Messages;
        this.newL1ToL2MessagesTreeRootSiblingPath = newL1ToL2MessagesTreeRootSiblingPath;
        this.startL1ToL2MessagesTreeSnapshot = startL1ToL2MessagesTreeSnapshot;
        this.startArchiveSnapshot = startArchiveSnapshot;
        this.newArchiveSiblingPath = newArchiveSiblingPath;
    }
    toBuffer() {
        return serializeToBuffer(...RootRollupInputs.getFields(this));
    }
    static from(fields) {
        return new RootRollupInputs(...RootRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.previousRollupData,
            fields.newL1ToL2Messages,
            fields.newL1ToL2MessagesTreeRootSiblingPath,
            fields.startL1ToL2MessagesTreeSnapshot,
            fields.startArchiveSnapshot,
            fields.newArchiveSiblingPath,
        ];
    }
}
/**
 * Represents public inputs of the root rollup circuit.
 *
 * NOTE: in practice, we'll hash all of this up into a single public input, for cheap on-chain verification.
 */
export class RootRollupPublicInputs {
    constructor(
    /**
     * Native aggregation state at the end of the rollup.
     */
    endAggregationObject, 
    /**
     * Global variables of the L2 block.
     */
    globalVariables, 
    /**
     * Snapshot of the note hash tree at the start of the rollup.
     */
    startNoteHashTreeSnapshot, 
    /**
     * Snapshot of the note hash tree at the end of the rollup.
     */
    endNoteHashTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the start of the rollup.
     */
    startNullifierTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the end of the rollup.
     */
    endNullifierTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the start of the rollup.
     */
    startContractTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the end of the rollup.
     */
    endContractTreeSnapshot, 
    /**
     * Root of the public data tree at the start of the rollup.
     */
    startPublicDataTreeRoot, 
    /**
     * Root of the public data tree at the end of the rollup.
     */
    endPublicDataTreeRoot, 
    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    startL1ToL2MessagesTreeSnapshot, 
    /**
     * Snapshot of the L1 to L2 message tree at the end of the rollup.
     */
    endL1ToL2MessagesTreeSnapshot, 
    /**
     * Snapshot of the blocks tree roots tree at the start of the rollup.
     */
    startArchiveSnapshot, 
    /**
     * Snapshot of the blocks tree roots tree at the end of the rollup.
     */
    endArchiveSnapshot, 
    /**
     * Hash of the calldata.
     */
    calldataHash, 
    /**
     * Hash of the L1 to L2 messages.
     */
    l1ToL2MessagesHash) {
        this.endAggregationObject = endAggregationObject;
        this.globalVariables = globalVariables;
        this.startNoteHashTreeSnapshot = startNoteHashTreeSnapshot;
        this.endNoteHashTreeSnapshot = endNoteHashTreeSnapshot;
        this.startNullifierTreeSnapshot = startNullifierTreeSnapshot;
        this.endNullifierTreeSnapshot = endNullifierTreeSnapshot;
        this.startContractTreeSnapshot = startContractTreeSnapshot;
        this.endContractTreeSnapshot = endContractTreeSnapshot;
        this.startPublicDataTreeRoot = startPublicDataTreeRoot;
        this.endPublicDataTreeRoot = endPublicDataTreeRoot;
        this.startL1ToL2MessagesTreeSnapshot = startL1ToL2MessagesTreeSnapshot;
        this.endL1ToL2MessagesTreeSnapshot = endL1ToL2MessagesTreeSnapshot;
        this.startArchiveSnapshot = startArchiveSnapshot;
        this.endArchiveSnapshot = endArchiveSnapshot;
        this.calldataHash = calldataHash;
        this.l1ToL2MessagesHash = l1ToL2MessagesHash;
    }
    static getFields(fields) {
        return [
            fields.endAggregationObject,
            fields.globalVariables,
            fields.startNoteHashTreeSnapshot,
            fields.endNoteHashTreeSnapshot,
            fields.startNullifierTreeSnapshot,
            fields.endNullifierTreeSnapshot,
            fields.startContractTreeSnapshot,
            fields.endContractTreeSnapshot,
            fields.startPublicDataTreeRoot,
            fields.endPublicDataTreeRoot,
            fields.startL1ToL2MessagesTreeSnapshot,
            fields.endL1ToL2MessagesTreeSnapshot,
            fields.startArchiveSnapshot,
            fields.endArchiveSnapshot,
            fields.calldataHash,
            fields.l1ToL2MessagesHash,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...RootRollupPublicInputs.getFields(this));
    }
    static from(fields) {
        return new RootRollupPublicInputs(...RootRollupPublicInputs.getFields(fields));
    }
    /**
     * Returns the sha256 hash of the calldata.
     * @returns The sha256 hash of the calldata.
     */
    sha256CalldataHash() {
        const high = this.calldataHash[0].toBuffer();
        const low = this.calldataHash[1].toBuffer();
        const hash = Buffer.alloc(32);
        for (let i = 0; i < 16; i++) {
            hash[i] = high[i + 16];
            hash[i + 16] = low[i + 16];
        }
        return hash;
    }
    /**
     * Deserializes a buffer into a `RootRollupPublicInputs` object.
     * @param buffer - The buffer to deserialize.
     * @returns The deserialized `RootRollupPublicInputs` object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new RootRollupPublicInputs(reader.readObject(AggregationObject), reader.readObject(GlobalVariables), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), [Fr.fromBuffer(reader), Fr.fromBuffer(reader)], [Fr.fromBuffer(reader), Fr.fromBuffer(reader)]);
    }
}
